src/Raven.Server/Documents/Sharding/Background/ShardedDocumentsMigrator.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Server.ServerWide.Commands.Sharding;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Logging;
+
+namespace Raven.Server.Documents.Sharding.Background
+{
+    public class ShardedDocumentsMigrator : IDisposable
+    {
+        private readonly ShardedDocumentDatabase _database;
+        private readonly CancellationToken _token;
+        private readonly Logger _logger;
+
+        public ShardedDocumentsMigrator(ShardedDocumentDatabase database)
+        {
+            _database = database;
+            _token = database.ShardedDocumentsStorage.DocumentDatabase.DatabaseShutdown;
+            _logger = LoggingSource.Instance.GetLogger(database.Name, GetType().FullName);
+        }
+
+        internal async Task ExecuteMoveDocumentsAsync()
+        {
+            try
+            {
+                int bucket = -1;
+                int moveToShard = -1;
+                bool found = false;
+                bool monitorTaken = false;
+                try
+                {
+                    _token.ThrowIfCancellationRequested();
+
+                    Monitor.TryEnter(this, 250, ref monitorTaken);
+                    if (monitorTaken == false)
+                        return;
+
+                    if (_database.ServerStore.Sharding.HasActiveMigrations(_database.ShardedDatabaseName))
+                        return;
+
+                    using (_database.ShardedDocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
+                    using (context.OpenReadTransaction())
+                    {
+                        var configuration = _database.ShardingConfiguration;
```

Comment:
Is this thread safe to do? 
Are we replacing this instance, or mutating it?

Reply:
yes, we are not changing `_database.ShardingConfiguration`/`configuration`